include/mqtt/property.hpp

```diff
@@ -99,18 +99,23 @@ struct n_bytes_property {
     boost::container::static_vector<char, N> buf_;
 };
 
-struct binary_property_ref;
-
 struct binary_property {
-    binary_property(property::id id, std::string str)
+    template <
```

Comment:
Why support template types? Not saying you're wrong, just curious what your reasoning is.

Reply:
This is the reason to use template.
https://wandbox.org/permlink/IJ4ZIPlN0ieD5rLO

Reply:
I noticed that it is not a good example. I will post other one later.

Reply:
This is actual case:
https://wandbox.org/permlink/8PyZa2JKwUgZM3Ak 

And I noticed that adding constructor `mqtt::string_view` can eliminate templates:
https://wandbox.org/permlink/uu06bJtVQn4C760b

Reply:
Hrmm.

I don't think its the best idea to have a constructor for MQTT::string_view.

My reason is that users of the code may assume that MQTT::buffer copies the string view, and therefore keeps the data alive.

Or, they may assume the opposite. Either way, we would have people assume the wrong thing.


Is your concern about string literals?

What about a user defined literal?

User defined literals have the special property that their lifetime is a superset of the programs lifetime.

We could do

```
constexpr MQTT::buffer operator "" _buf(const char* str, size_t len) noexcept
{
    // return MQTT::buffer constructed without copying the string data, and a nullptr lifetime
}

```

Reply:
That would get used like:

`binary_property(1, "ABC"_buf);`

Reply:
I understand why `mqtt::string_view` overload is dangerous. I will implement using user defined literals.

Reply:
Fixed.